flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedSql.java

```diff
@@ -77,4 +100,92 @@ protected List<Row> run(
     String optionStr = SqlHelpers.sqlOptionsToString(options);
     return SqlHelpers.sql(getTableEnv(), "select %s from t %s %s", select, optionStr, sqlFilter);
   }
+
+  protected Record generateRecord(Instant t1, long t2) {
+    Record record = GenericRecord.create(SCHEMA_TS);
+    record.setField("t1", t1.atZone(ZoneId.systemDefault()).toLocalDateTime());
+    record.setField("t2", t2);
+    return record;
+  }
+
+  @Test
+  public void testWatermarkOptions() throws Exception {
+    // Skip AVRO since we don't collect metrics for it, and hence we cannot use watermark column as
+    // there are no stats
+    // re: https://github.com/apache/iceberg/pull/1963
+    if (fileFormat != FileFormat.AVRO) {
+      Table table = catalogResource.catalog().createTable(TestFixtures.TABLE_IDENTIFIER, SCHEMA_TS);
+      long baseTime = 1702382109000L;
+
+      GenericAppenderHelper helper = new GenericAppenderHelper(table, fileFormat, TEMPORARY_FOLDER);
+
+      // File 1 - early timestamps, new longs
+      Record early1 =
+          generateRecord(Instant.ofEpochMilli(baseTime), baseTime + (1000 * 60 * 60 * 24 * 30L));
+      Record early2 =
+          generateRecord(
+              Instant.ofEpochMilli(baseTime - 10 * 1000L), baseTime + (1000 * 60 * 60 * 24 * 35L));
+
+      List<Record> recordsDataFile1 = Lists.newArrayList();
+      recordsDataFile1.add(early1);
+      recordsDataFile1.add(early2);
+      DataFile dataFile1 = helper.writeFile(recordsDataFile1);
+      // File 2 - old timestamps, old longs
+      Record late1 =
+          generateRecord(
+              Instant.ofEpochMilli(baseTime + 14 * 1000L), baseTime - (1000 * 60 * 60 * 24 * 30L));
+      Record late2 =
+          generateRecord(
+              Instant.ofEpochMilli(baseTime + 12 * 1000L), baseTime - (1000 * 60 * 61 * 24 * 35L));
+
+      List<Record> recordsDataFile2 = Lists.newArrayList();
+      recordsDataFile2.add(late1);
+      recordsDataFile2.add(late2);
+      DataFile dataFile2 = helper.writeFile(recordsDataFile2);
+
+      helper.appendToTable(dataFile1, dataFile2);
+
+      // first assertion: early and then late
+      List<Record> expected = Lists.newArrayList();
+      expected.addAll(recordsDataFile1);
+      expected.addAll(recordsDataFile2);
+      TestHelpers.assertRecords(
```

Comment:
Is this test fails, if you change the order of the records?
I expect that `TestHelpers.assertRecords` is independent of the order. See: https://github.com/apache/iceberg/blob/22b95dc70eb46e82a0d51fd4cfb71ad4cbdf946a/flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestHelpers.java#L146